src/datasets/utils/download_manager.py

```diff
@@ -225,21 +247,37 @@ def iter_archive(self, path):
             Generator yielding tuple (path_within_archive, file_obj).
             File-Obj are opened in byte mode (io.BufferedReader)
         """
-        with open(path, "rb") as f:
-            stream = tarfile.open(fileobj=f, mode="r|*")
-            for tarinfo in stream:
-                file_path = tarinfo.name
-                if not tarinfo.isreg():
+        extension = _get_extraction_protocol_local(path)
+
+        if extension == "tar":
+            with open(path, "rb") as f:
+                stream = tarfile.open(fileobj=f, mode="r|*")
+                for tarinfo in stream:
+                    file_path = tarinfo.name
+                    if not tarinfo.isreg():
+                        continue
+                    if file_path is None:
+                        continue
+                    if os.path.basename(file_path).startswith(".") or os.path.basename(file_path).startswith("__"):
+                        # skipping hidden files
+                        continue
+                    file_obj = stream.extractfile(tarinfo)
+                    yield (file_path, file_obj)
+                    stream.members = []
+                del stream
+
+        elif extension in ["zip", "gzip"]:
+            zipf = zipfile.ZipFile(path)
+            for member in zipf.infolist():
+                file_path = member.filename
+                if member.is_dir():
                     continue
                 if file_path is None:
                     continue
                 if os.path.basename(file_path).startswith(".") or os.path.basename(file_path).startswith("__"):
-                    # skipping hidden files
                     continue
-                file_obj = stream.extractfile(tarinfo)
+                file_obj = io.BufferedReader(zipf.open(member, "r"))
                 yield (file_path, file_obj)
```

Comment:
Let's raise an error if it's not TAR or ZIP, just in case:
```suggestion
                yield (file_path, file_obj)
            else:
                raise NotImplementedError(f"DownloadManager.iter_archive isn't implemented for '{extension}' file at {path}")
```